fix: change `zod` schemas inputs/outputs to `readonly` to match abitype

mathieu-bour opened the pull request:
## Description

This PR upgrades [`zod` to v3.22+ to get access to `ZodReadonly`](https://github.com/colinhacks/zod/releases/tag/v3.22.0).
This allow to update the `zod` schemas from `abitype/zod` so  type from `abitype`.

Added readonly to the following types:

- `AbiContructor.inputs`
- `AbiError.inputs`
- `AbiEvent.inputs`
- `AbiFunction.inputs` / `AbiFunction.outputs`

## Additional Information

- [X] I read the [contributing guide](https://github.com/wagmi-dev/abitype/blob/main/.github/CONTRIBUTING.md)
- [X] I added documentation related to the changes made.
- [X] I added or updated tests related to the changes made.

Your ENS/address: `windyy.eth`

<!-- start pr-codex -->

---

## PR-Codex overview
### Detailed summary
- Changed the following types to readonly in zod package: `AbiContructor.inputs`, `AbiError.inputs`, `AbiEvent.inputs`, `AbiFunction.inputs` / `AbiFunction.outputs`
- Updated the `zod` package version to `3.22.4` in `pnpm-lock.yaml`
- Updated the `zod` package version to `^3 >=3.22.0` in `package.json`
- Added a note in `CONTRIBUTING.md` to use `pnpm@8.3.1` and `typescript@5.0.4` for installing or upgrading packages
- Added a note in `CONTRIBUTING.md` to build the package before running the `test:typecheck` suite
- Updated the `zod.ts` file to make the `components`, `inputs`, and `outputs` properties readonly in various types
- Updated the `zod.test-d.ts` file to import additional types and test their assignability and extension

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

Raiden1411 commented:
In order to run the `pnpm test:typecheck` you need to run `pnpm build` first. You can take a look at how the CI runs on tests to get a better idea of it.
I think it would also make sense in adding the `readonly` property to `AbiParameter` on where ever it gets called. Would you mind adding it?

mathieu-bour commented:
@Raiden1411 Thanks a lot for clarifications, I was able to test the types as expected. I added a line in `.github/CONTRIBUTING.md` about this specific case.

Moreover, I added all relevant .readonly() to `inputs` and `outputs`.

Raiden1411 commented on .github/CONTRIBUTING.md:
````diff
@@ -95,6 +95,8 @@ pnpm test
 pnpm test:typecheck
 ```
 
+Note: ensure to build the package (`pnpm build`) before running the `test:typecheck` suite.
````
Lets change this from `Note: ensure` to `> **Note** Ensure`

Raiden1411 commented on package.json:
```diff
@@ -70,15 +70,9 @@
+      "zod": ["./dist/types/zod.d.ts"]
     }
   },
   "peerDependencies": {
```
Could you change the `zod` peer dependency here to match the new zod version?

Raiden1411 commented on .github/CONTRIBUTING.md:
````diff
@@ -95,6 +95,8 @@ pnpm test
 pnpm test:typecheck
 ```
 
+Note: ensure to build the package (`pnpm build`) before running the `test:typecheck` suite.
````
Lets change this from Note: ensure to > **Note** Ensure.

Also can you add a section in the `Install` section to make sure to use typescript version `5.0.4`

mathieu-bour replied:
Done in 238b89e, but a bit unsure about the change

Raiden1411 commented on pnpm-lock.yaml:
```diff
@@ -1,5 +1,9 @@
 lockfileVersion: '6.0'
 
+settings:
```
Are you using `pnpm@8.3.1`? If not could you reinstall with it? Also make sure to use typescript version `5.0.4` as it the one currently used by Abitype.

mathieu-bour replied:
fixed in 24dd673

Raiden1411 reviewed (changes requested):
Needs final suggested changes

Raiden1411 commented on .github/CONTRIBUTING.md:
```diff
@@ -54,13 +54,14 @@ gh repo clone wagmi-dev/abitype
 ## Installing Node.js and pnpm
 
-ABIType uses [pnpm](https://pnpm.io) as its package manager. You need to install **Node.js v16 or higher** and **pnpm v7 or higher**.
+ABIType uses [pnpm](https://pnpm.io) as its package manager. You need to install **Node.js v16 or higher**, **pnpm exactly v8.3.1** and **TypeScript exactly v5.0.4**.
```
Sorry my wording in the previous comment was incorrect. I was more thinking in adding a "Note" for anyone that might want to install packages. Something like this.

> **Note:** In case you have to install new packages or upgrade packages make sure to use **pnpm@8.3.1** and **typescript@5.0.4**

You can revert the message to the older one and add this note

mathieu-bour replied:
Okay, no problem!
